java-cfenv-boot-pivotal-sso/src/main/java/io/pivotal/cfenv/boot/sso/CfSingleSignOnProcessor.java

```diff
@@ -46,12 +57,36 @@ public void process(CfCredentials cfCredentials, Map<String, Object> properties)
-        properties.put(SPRING_SECURITY_CLIENT + ".registration.sso.client-secret", clientSecret);
-        properties.put(SPRING_SECURITY_CLIENT + ".registration.sso.client-name", SSO_CLIENT_NAME);
-        properties.put(SPRING_SECURITY_CLIENT + ".registration.sso.redirect-uri", "{baseUrl}/login/oauth2/code/{registrationId}");
-        properties.put(SPRING_SECURITY_CLIENT + ".provider.sso.issuer-uri", issuer + "/oauth/token");
-        properties.put(SPRING_SECURITY_CLIENT + ".provider.sso.authorization-uri", authDomain + "/oauth/authorize");
+        properties.put(SPRING_SECURITY_CLIENT + ".provider." + PROVIDER_ID + ".issuer-uri", issuer + "/oauth/token");
+        properties.put(SPRING_SECURITY_CLIENT + ".provider." + PROVIDER_ID + ".authorization-uri", authDomain + "/oauth/authorize");
+
+        ArrayList<String> grantTypes = (ArrayList<String>) cfCredentials.getMap().get("grant_types");
+        if (grantTypes != null && isAuthCodeAndClientCreds(grantTypes)) {
+            mapBasicClientProperties(properties, AUTHCODE_CLIENT_REGISTRATION_ID, clientId, clientSecret);
+            properties.put(SPRING_SECURITY_CLIENT + ".registration." + AUTHCODE_CLIENT_REGISTRATION_ID + ".authorization-grant-type", AUTHORIZATION_CODE);
+
+            mapBasicClientProperties(properties, CLIENTCRED_CLIENT_REGISTRATION_ID, clientId, clientSecret);
+            properties.put(SPRING_SECURITY_CLIENT + ".registration." + CLIENTCRED_CLIENT_REGISTRATION_ID + ".authorization-grant-type", CLIENT_CREDENTIALS);
+        } else if (grantTypes != null && grantTypes.size() == 1) { // if one grant type
+            mapBasicClientProperties(properties, BASE_CLIENT_REGISTRATION_ID, clientId, clientSecret);
+            String grantType = grantTypes.get(0);
+            properties.put(SPRING_SECURITY_CLIENT + ".registration." + BASE_CLIENT_REGISTRATION_ID + ".authorization-grant-type", grantType);
+        } else { // if grant type is empty, invalid combo, or more than 2 grant types
+            mapBasicClientProperties(properties, BASE_CLIENT_REGISTRATION_ID, clientId, clientSecret);
+        }
+    }
+
+    private boolean isAuthCodeAndClientCreds(ArrayList<String> grantTypes) {
+        return (new HashSet<>(grantTypes)).equals(AUTH_CODE_AND_CLIENT_CREDS);
+    }
+
+    private void mapBasicClientProperties(Map<String, Object> properties, String clientRegistrationId, String clientId, String clientSecret) {
+        String registrationPrefix = SPRING_SECURITY_CLIENT + ".registration." + clientRegistrationId;
+        properties.put(registrationPrefix + ".client-id", clientId);
+        properties.put(registrationPrefix + ".client-secret", clientSecret);
+        properties.put(registrationPrefix + ".client-name", clientRegistrationId);
+        properties.put(registrationPrefix + ".redirect-uri", "{baseUrl}/login/oauth2/code/{registrationId}");
+        properties.put(registrationPrefix + ".provider", PROVIDER_ID);
```

Comment:
This looks like a new property, is wasn't required in the past?

Reply:
The grant type presence should not matter for this property. 

With this change set, the provider is declared separately under the a name independent of the grant type. The registration(s) now refers to this provider by this hardcoded name. This slack thread may be helpful context: https://pivotal.slack.com/archives/C06F3S60H/p1564439547048000

Reply:
I see, ok, thanks.